packages/react-dom/src/__tests__/ReactMount-test.js

```diff
@@ -173,12 +482,632 @@ describe('ReactMount', () => {
+      <div>
+        <div data-ssr-mismatch-padding-before="1" />
+        <TestPaddingBeforeComponent />
+        <div data-ssr-mismatch-padding-before="10" />
+        <div data-ssr-mismatch-padding-before="11" />
+        <div data-ssr-mismatch-padding-before="12" />
+      </div>,
+    );
+    div.innerHTML = markup;
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div>
+          <div data-ssr-mismatch-padding-before="1" />
+          <TestPaddingBeforeComponent />
+          <div data-ssr-mismatch-padding-before="10" />
+          <div data-ssr-mismatch-padding-before="11" />
+          <div data-ssr-mismatch-padding-before="12" />
+          SSRMismatchTest client text
+        </div>,
+        div,
+      ),
+    ).toWarnDev(
+      'Warning: Expected server HTML to contain a matching text node' +
+        " for {'SSRMismatchTest client text'} in <div>.\n\n" +
+        '  <div data-reactroot="">\n' +
+        '    <div data-ssr-mismatch-padding-before="1" />\n' +
+        '    <div data-ssr-mismatch-padding-before="2" />\n' +
+        '    <div data-ssr-mismatch-padding-before="3" />\n' +
+        '    <div data-ssr-mismatch-padding-before="4" />\n' +
+        '    <div data-ssr-mismatch-padding-before="5" />\n' +
+        '    <div data-ssr-mismatch-padding-before="6" />\n' +
+        '    <div data-ssr-mismatch-padding-before="7" />\n' +
+        '    <div data-ssr-mismatch-padding-before="8" />\n' +
+        '    <div data-ssr-mismatch-padding-before="9" />\n' +
+        '    <div data-ssr-mismatch-padding-before="10" />\n' +
+        '    <div data-ssr-mismatch-padding-before="11" />\n' +
+        '    <div data-ssr-mismatch-padding-before="12" />\n' +
+        "+   {'SSRMismatchTest client text'}\n" +
+        '  </div>\n\n' +
+        '    in div (at **)',
+    );
+  });
+
+  it('should warn when hydrate inserts an element after a comment node (insertion diff)', () => {
+    const div = document.createElement('div');
+    div.innerHTML = '<div><!-- a comment --></div>';
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div>
+          <span />
+        </div>,
+        div,
+      ),
+    ).toWarnDev(
+      'Warning: Expected server HTML to contain a matching <span> in <div>.\n\n' +
+        '  <div>\n' +
+        '    <!-- a comment -->\n' +
+        '+   <span />\n' +
+        '  </div>\n\n' +
+        '    in span (at **)\n' +
+        '    in div (at **)',
+    );
+  });
+
+  it('should warn when hydrate replaces an element after a comment node (replacement diff)', () => {
+    const div = document.createElement('div');
+    div.innerHTML = '<div><!-- a comment <-- --&gt; > --><div></div></div>';
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div>
+          <span />
+          <div />
+        </div>,
+        div,
+      ),
+    ).toWarnDev(
+      'Warning: Expected server HTML to contain a matching <span> in <div>.\n\n' +
+        '  <div>\n' +
+        '    <!-- a comment <-- --&gt; > -->\n' +
```

Comment:
Here I test that the HTML comment is printed the same way it would've been written in the original markup, `&gt;` ensures the HTML comment is not ended prematurely.